src/main/kotlin/io/kuzzle/sdk/coreClasses/http/Route.kt

```diff
@@ -0,0 +1,178 @@
+package io.kuzzle.sdk.coreClasses.http
+
+import io.kuzzle.sdk.coreClasses.exceptions.MissingURLParamException
+import io.kuzzle.sdk.coreClasses.json.JsonSerializer
+import io.kuzzle.sdk.coreClasses.maps.KuzzleMap
+import java.net.URLEncoder
+
+enum class PartType {
+    STATIC,
+    TEMPLATE
+}
+
+private class RoutePart {
+    val type: PartType
+    val value: String
+
+    constructor(type: PartType, value: String) {
+        this.type = type
+        this.value = value
+    }
+}
+
+val PayloadProperties = setOf<String>(
+    "controller",
+    "action",
+    "meta",
+    "volatile",
+)
+
+class Route {
+    private val partType: PartType
+    private val routeParts: ArrayList<RoutePart>
+    private var staticURL: String = ""
+    private var baseURL: String
+    private var verb: String
+
+    private constructor(verb: String, baseUrl: String, routeParts: ArrayList<RoutePart>) {
+        this.verb = verb.uppercase()
+        this.baseURL = baseUrl
+        this.routeParts = routeParts
+        val hasOnePartTemplated = routeParts.indexOfFirst { it.type == PartType.TEMPLATE } > -1
+        this.partType = if (hasOnePartTemplated) PartType.TEMPLATE else PartType.STATIC
+        if (! hasOnePartTemplated) {
+            staticURL = routeParts.joinToString("") {
+                it.value
+            }
+        }
+    }
+
+    /**
+     * Given map containing the properties of the request, the buildRequest method
+     * will create an instance of HttpRequest which contains the information needed to make
+     * the request later on.
+     *
+     * The method has to:
+     * - Generate the URL using the properties of the request if there are part of the URL that are templated
+     * - Generate the query string
+     * - Generate the list of headers
+     */
+    fun buildRequest(request: KuzzleMap): HttpRequest {
+        val headers = request.optMap("headers", KuzzleMap())
+
+        val queryArgs = KuzzleMap()
+        for (key: String? in request.keys) {
+            // Skip if a key or value is null
+            if (key == null || request[key] == null) {
+                continue
+            }
+
+            when (key) {
+                "jwt" -> headers["authorization"] = "Bearer ${request["jwt"]}"
+                "volatile" -> {
+                    headers["x-kuzzle-volatile"] = request["volatile"]
+                }
+                "headers" -> headers.putAll(request.optMap("headers", KuzzleMap()))
+                "body" -> {

```

Comment:
I was wondering too, but I simply copied how the SDK Javascript behaves

Reply:
We need to keep this as it his because controller are making request using the body property and if the HTTP Endpoint used is a GET we need to transfer the body properties into the query string otherwise information will be lost.

Reply:
Ok I understand

Reply:
I still don't understand since when you execute an API action you put parameter either at the root level or in the `body` object and it is not handled by Kuzzle in the same way when using the WebSocket protocol so why should it be for the HTTP?

Reply:
Request using the Websocket protocol are sent with an object like this:

```js
// Websocket version
sdk.query({
  controller: 'document',
  action: 'create',
  index: 'my-index',
  collection: 'my-collection',
  body: {
    fieldA: 'value',
    fieldB: 'value',
  }
});
```

If we are not putting the body parameters inside the query parameters automatically it would means that we have to do it manually like so
```js
// HTTP Version
sdk.query({
  controller: 'document',
  action: 'create',
  index: 'my-index',
  collection: 'my-collection',
  fieldA: 'value',
  fieldB: 'value',
});
```

The second version will also work but since the controller are calling query also for the Websocket we need to either duplicate code to call query from the controller using the first way or the second way of writing queries based on the protocol used (Websocket or HTTP) for each controller's action, or  we can let the HTTP Protocol do that by himself which is exactly the same.

Reply:
:+1: